plugin/src/compilation/decompiler.ml

```diff
@@ -70,63 +101,101 @@ let get_pushed_names env =
             failwith "Unexpected Anonymous in get_pushed_names."
          | Name n -> n) names)
 
+(* Checks if we are applying function g, which
+   is expected to have len arguments. *)
+let try_fun f args g len : (constr array) option =
+  if Array.length args == len && equal f g then
+    Some args
+  else
+    None
+  
 (* A proper rewrite must have 6 arguments and be one of the
    common eliminators for equality generated by coq. *)
-let try_rewrite f args rewr default =
+let try_rewrite f args : (bool * constr * constr) option =
   if Array.length args == 6 && is_rewrite f then
     let left = is_rewrite_l f in
     let prf = Array.get args 5 in
     let hyp = Array.get args 3 in
-    rewr left prf hyp
+    Some (left, prf, hyp)
   else
-    default
+    None
+
+(* Option monad over function application. *)
+let try_app (trm : constr) : (constr * constr array) option =
+  match kind trm with
+  | App (f, args) -> Some (f, args)
+  | _ -> None
+
+(* Option monad over relative indices. *)
+let try_rel (trm : constr) : int option =
+  match kind trm with
+  | Rel i -> Some i
+  | _ -> None
+
+(* Guard fails if the rel index occurs in the term. *)
+let noccur (i : int) (trm : constr) : unit option =
+  if Vars.noccurn i trm then Some () else None
+
+let rec first_pass env trm =
```

Comment:
Can you add a comment above this that explains why you need it, briefly? Perhaps point to an example

Reply:
I see this is from earlier, so I should have caught this earlier, but the comment would still be useful

Reply:
`try_app` and `try_rel` are necessary for rewrite-in/apply-in/induction. Using rewrite-in as the example:
```
(* we have let _ := valu in body *)
let valu = Reduction.whd_betaiota env valu in      
try_app valu          >>= fun (f, args) ->         (* let _ := "f args" in _ *)
try_rewrite f args    >>= fun (left, prf, hyp) ->  (* "f args" == "eq_ind ... prf ... hyp ..." *)
try_rel hyp           >>= fun idx ->               (* "hyp" refers to a hypothesis in the context (not an expression) *)
noccur (idx + 1) body >>= fun _ ->                 (* "idx" is never used in the body of the let-expression *)
```
noccur is a guard, so it doesn't compute anything but it will break execution if the test fails, considering making a more general  `bool -> unit option` function instead.

Reply:
Here I meant `first_pass` (follow-up from IRL conversation)